### ISSUE TITLE
Question (issue?) about Dependency Submission and GitHub Dependabot

### ISSUE THREAD

**Issue by nnellanspdl** [gradle/actions#80]
Can you tell me if this is how things are supposed to work?

I have a Gradle project hosted in a GitHub repo.  I have a simple GitHub workflow that runs a standard `./gradlew build` and also uses the dependency-submission action.  My project has a `build.gradle` file with quite a few things listed directly under `dependencies{}`.

I would like to make use of GitHub Dependabot **Security** Updates.  This should automatically open PRs for any dependencies that have vulnerabilities.  However, when I look at my repo's dependencies, (which were created and uploaded by the action), then I only see transitive dependencies, which all show as coming from the `settings.gradle` file.

I should be seeing my direct dependencies coming from my `build.gradle` file, right?  Since these are showing as indirect, and from `settings.gradle` then Dependabot Security Updates can't create an automatic PR for them:  `This is an indirect dependency not explicitly present in any manifest file, so Dependabot cannot update it.`

Is there anything I can do about this?

**Comment by nnellanspdl** [comment]
I'm not sure I understand your response.

So what you're saying is the direct dependencies that I have listed in my `build.gradle` file will NOT be picked up by Dependabot Security Updates?

I thought this was supposed to work?  https://github.blog/changelog/2023-08-24-gradle-support-for-dependabot-security-updates/

**Comment by bigdaz** [comment]
> So what you're saying is the direct dependencies that I have listed in my build.gradle file will NOT be picked up by Dependabot Security Updates? This is currently what is happening, I have 20+ dependencies listed in my build.gradle file, but Security Updates doesn't pick these up as direct dependencies.

The dependency graph is not used when Dependabot attempts to create a PR to update a dependency version. Instead, various heuristics are used to attempt to find where these dependencies are declared. I would expect that dependencies declared directly in a `build.gradle` file would be detected.

Are you able to share any specifics of your project? 
- How are the dependencies declared? Why do you think they are all flagged as "indirect" dependencies in your dependency graph?
- Are you able to provide a simple example project where I can reproduce this issue? A dependency version that is declared in your build.gradle file should be considered "direct" in your dependency graph (even though it's attributed to 'settings.gradle').

**Comment by nnellanspdl** [comment]
Thanks for the response.  This is a private repo, so unfortunately I can't share it.  But, I will do my best to explain below.

Here are some snippets from my `build.gradle` file.  This is all the plugins I'm using, however I have a lot more dependencies than what I'm listing here, I'm truncating it to save space:
```
plugins {
	id 'org.springframework.boot' version '2.3.5.RELEASE'
	id 'io.spring.dependency-management' version '1.0.10.RELEASE'
	id 'java'
	id 'jacoco'
	id "org.sonarqube" version "4.3.0.3225"
}
dependencies {
	implementation "org.projectlombok:lombok:1.18.22"
	implementation "org.springframework.boot:spring-boot-starter-web:2.3.5.RELEASE"
	implementation "org.springframework.boot:spring-boot-starter-actuator:2.3.5.RELEASE"
	implementation "com.azure:azure-spring-data-cosmos:3.3.0"
	implementation "com.google.code.gson:gson:2.8.6"
}
```

I know for a fact that the `gson` dependency has a vulnerability that I'm affected by: [source](https://github.com/advisories/GHSA-4jrv-ppp4-jm57).  So, I will use this one as an example moving forward.

Next, I use a workflow to do `gradlew build` and run the dependency submission action.  I'm using the action's output `dependency-graph-file` to upload that dependency file as a build artifact.  When I look at the file I can see the following:
```json
"com.google.code.gson:gson:2.8.6" : {
    "package_url" : "pkg:maven/com.google.code.gson/gson@2.8.6",
    "relationship" : "direct",
    "dependencies" : [ ]
},
```
That tells me the action is picking it up as a "direct" dependency, right?

Repo > Insights > Dependency Graph > Dependencies > search for `gson`
* I see this dependency:  `com.google.code.gson:gson 2.8.6`
* It shows as having 1 "high risk" vulnerability
* As you've already stated, it shows this dependency came from `settings.gradle` even though it came from `build.gradle`

Repo > Security > Dependabot (this is Dependabot Alerts)
* I see the same thing as above.  It has a "high" alert for this dependency, and it shows as coming from `settings.gradle`
* Dependabot Security Updates ARE enabled on my repo.  However, no PRs are automatically opened for this alert.  This is confusing because of Microsoft's [announcement](https://github.blog/changelog/2023-08-24-gradle-support-for-dependabot-security-updates/).  They clearly say that "If you have used the dependency submission API to upload your Gradle dependencies to the dependency graph and are receiving Dependabot alerts for those dependencies, Dependabot will now try to open a pull request to resolve them automatically if you have security updates enabled for your repository."
* Dependabot Alerts gives you a button which tries to manually open a PR for this alert.  When I try this I get the previously mentioned error:  "This is an indirect dependency not explicitly present in any manifest file, so Dependabot cannot update it."

Dependabot Version Updates
* Next, I enabled Dependabot Version Updates by creating a `.gihub/dependabot.yml` file with the following contents.
```yaml
version: 2
updates:
  - package-ecosystem: "gradle"
    directory: "/"
    schedule:
      interval: "weekly"
```
* This works, and I get new PRs automatically opened.


Hopefully that wasn't too confusing.  I appreciate the help here.

**Comment by bigdaz** [comment]
OK, so it looks like Dependabot Security Updates are somehow using a different algorithm than Dependabot Version Updates when constructing PRs. I don't know why this would be different, but unfortunately it appears that Dependabot Security Updates are not working well with the `dependency-submission` action.

It's possible that if you changed the dependency graph to `build.gradle` and re-submitted it directly, then these particular Security Updates would work. However, it's not possible for the `dependency-submission` action to accurately map each dependency version to a specific "manifest file" for 2 reasons:

1. The Gradle events that are received when resolving dependencies don't contain this information. It may be possible (but difficult), to develop a plugin that extracts this information.
2. The GitHub Dependency Submission API is not suitable for the Gradle model where: a) the same dependency can be declared and consumed in many build files and b) many files (including plugins) can contribute logic used to determine dependency versions.

We have long-term plans to improve the first point, and we are working with GitHub to resolve the second. However, at this stage the behaviour your are experiencing is what is expected. I recommend that you file an issue with Dependabot pointing out that the _version_ updates are able to create a PR where the _security_ updates are not: this is unexpected.

**Comment by nnellanspdl** [comment]
I did a quick test.  I updated my workflow to insert a quick bash command to modify the `dependency-graph-file` and replace `settings.gradle` with `build.gradle`

The submission worked okay.  Dependabot Alerts closed all of the old `settings.gradle` alerts and opened the same exact number of new alerts but this time they were showing `build.gradle.

Unfortunately, Security PRs were not automatically opened.  And when I tried to manually create a Security PR it gave me the same error about it being an indirect dependency.

**Comment by bigdaz** [comment]
I created [this demo project](https://github.com/bigdaz/dependency-submission-demo) and was able to successfully generate a [Dependabot Security Update PR](https://github.com/bigdaz/dependency-submission-demo/pull/1) by clicking on the appropriate button.

<img width="982" alt="image" src="https://github.com/gradle/actions/assets/179734/49911e71-6ae8-4c24-99bb-d4430c6ef487">

You'll note that the message says it's updating a dependency in `settings.gradle.kts`, but the [actual PR updates the version](https://github.com/bigdaz/dependency-submission-demo/pull/1/files) in `gradle/lib.versions.toml`.

**Comment by bigdaz** [comment]
Thanks for your questions, by the way. I'm planning to put together a FAQ covering dependency-submission.
#85 is a start.

**Comment by nnellanspdl** [comment]
That's a fantastic FAQ!  Would have saved me quite a lot of time.  But, I've also learned a lot along the way, so it wasn't wasted time.

I'm still going to play around with various combinations of moving my dependency config.  First, to a settings.gradle file, and then I'll try the lib.versions.toml file.

**Comment by nnellanspdl** [comment]
I almost hate to admit it, but I figured this out.

We defined our dependencies in our `build.gradle` as such:
```
dependencies {
  implementation "com.google.code.gson:gson:$gson_version"
}
```

And then we have a `gradle.properties` file where we define:
```
gson_version=2.8.6
```

Dependabot doesn't know how to handle this.  Dependabot only supports updating versions in your `build.gradle` if they are explicitly defined.  I took out the `$gson_version` variable and replaced it with an explicit version number, and then it worked!
